### DISCUSSION TITLE
composing 2 rgba images with a mask (but not use their alphas)

### DISCUSSION THREAD

**Discussion by sschoellhammer** [ImageMagick/ImageMagick#3454] [Help] [answered]
Hello there,

As the title says, I'm trying to compose 2 rgba images using a third mask image. The tricky bit is that the alpha channels of those 2 images don't contain "transparency" but arbitrary data, so it's not really composing in the traditional way. 

Most compose methods in magick use the alphas of the images ("over" and such). 
The only option I have found that works close to what I had in mind is "copy"
i.e:
magick blinn1_nms.tga blinn2_nms.tga blinn2_Mask.tga **-compose copy** -composite out.tga

Sadly with this method the alpha of the second image is simply cleared (the first one is maintained). 

So I'm really close to what I want and I wonder if someone has an idea to get me all the way there..

I have attached my test images here:
[compose_test.zip](https://github.com/ImageMagick/ImageMagick/files/6215486/compose_test.zip)

Thanks for the help!
seb

**Comment by snibgo** [comment]
> ... so it's not really composing in the traditional way.

Fair enough, but you don't say how you want the composition to work.

You might want to totally ignore all alpha values, so the output is opaque. If so, then use "-alpha off" after reading all the inputs.

Or you might want the output alpha to be a masked composite of the input alpha. If so, then "-alpha extract", then composite just the alpha, and "-compose CopyOpacity -composite" that into the opaque output.

Or you might want something different.

**Reply by sschoellhammer** [reply]
Sorry, I should have been more clear. 

What I essentially want is this:
rgba_result = rgba_img1 * mask + rgba_img1 * (1-mask)

So "-alpha off" seems to remove the alpha from the output image, which I don't want.
I'm not sure how to format the second command, would that be 2 separate steps, first saving out a temp image?

Thanks again for helping out!

**Comment by fmw42** [comment, marked as answer]
I think this might be what you want. You want the alpha channel to be processed just like the rgb channels.

```
magick rgba1 -alpha extract alpha1
magick rgba2 -alpha extract alpha2
magick alpha1 alpha2 mask -compose over -composite alpha12
magick rgba1 rgba2 mask -alpha off -compose over -composite rgb12
magick rgb12 alpha12 -alpha off -compose copy_opacity -composite result
```

The above could be combined into one command line using parenthesis processing. But I thought this would explain it better. Be sure to add your desired suffixes. Do not use JPG for intermediate processing results. Better to use PNG or TIFF for all results.

You might try also

`magick rgba1 rgba2 mask -channel rgba -compose over -composite result
`
I do not know if that will work or not. Not tested.

**Reply by sschoellhammer** [reply]
Thanks so much for taking the time to explain. 
The second command you mentioned did not work but I managed to clobber everything above into one big command as you suggested and this is doing the job!
Very much appreciated!
